base/src/main/java/proguard/optimize/inline/LambdaInliner.java

```diff
@@ -0,0 +1,103 @@
+package proguard.optimize.inline;
+
+//import com.guardsquare.gui.Gui;
+import proguard.optimize.inline.lambda_locator.LambdaLocator;
+import proguard.AppView;
+import proguard.classfile.AccessConstants;
+import proguard.classfile.Method;
+import proguard.classfile.ProgramClass;
+import proguard.classfile.editor.CodeAttributeEditor;
+import proguard.classfile.editor.ConstantPoolEditor;
+import proguard.classfile.instruction.ConstantInstruction;
+import proguard.classfile.instruction.Instruction;
+import proguard.classfile.util.InitializationUtil;
+import proguard.optimize.inline.lambda_locator.Lambda;
+import proguard.pass.Pass;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class LambdaInliner implements Pass {
+    private final String classNameFilter;
+    private boolean inlinedAllUsages;
+    public LambdaInliner(String classNameFilter) {
+        this.classNameFilter = classNameFilter;
+        this.inlinedAllUsages = true;
+    }
+
+    public LambdaInliner() {
+        this("");
+    }
+
+    @Override
+    public void execute(AppView appView) {
+        LambdaLocator lambdaLocator = new LambdaLocator(appView.programClassPool, classNameFilter);
+
+        for (Lambda lambda : lambdaLocator.getStaticLambdas()) {
+            Set<InstructionAtOffset> remainder = new HashSet<>();
+            inlinedAllUsages = true;
+            InitializationUtil.initialize(appView.programClassPool, appView.libraryClassPool);
+            lambda.codeAttribute().accept(lambda.clazz(), lambda.method(), new LambdaUsageFinder(lambda, lambdaLocator.getStaticLambdaMap(), appView, (foundLambda, consumingClass, consumingMethod, consumingCallOffset, consumingCallClazz, consumingCallmethod, consumingCallCodeAttribute, sourceTrace, possibleLambdas) -> {
+                // Try inlining the lambda in consumingMethod
+                BaseLambdaInliner baseLambdaInliner = new BaseLambdaInliner(appView, consumingClass, consumingMethod, lambda);
+                Method inlinedLambamethod = baseLambdaInliner.inline();
+
+                // We didn't inline anything so no need to change any call instructions.
+                if (inlinedLambamethod == null) {
+                    inlinedAllUsages = false;
+                    return;
+                }
+
+                if (possibleLambdas.size() > 1) {
+                    // This lambda is part of a collection of lambdas that might potentially be used, but we do not know which one is actually used. Because of that we cannot inline it.
+                    inlinedAllUsages = false;
+                    return;
+                }
+
+                CodeAttributeEditor codeAttributeEditor = new CodeAttributeEditor();
+                codeAttributeEditor.reset(consumingCallCodeAttribute.u4codeLength);
+
+                /*
+                 * Remove usages that bring the variable on the stack so all the way up until a load
```

Comment:
The lambda usage finder uses the partial evaluator to find lambda usages and this results in it automatically being smart enough to figure out something is comming from a variable or if it's just passed directly as an argument.

It is probably possible to disable inlining from variables by not allowing traces that make use of load and store operations. We currently kept this behaviour in place because it works fine and also because it's not really a case of just removing very specific code to handle variables.